File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = statement.trim match {
    case s if isQuestion(s) && isUpperCase(s) => "Calm down, I know what I'm doing!"
    case s if isUpperCase(s) => "Whoa, chill out!"
    case s if isQuestion(s) => "Sure."
    case s if isSilence(s) => "Fine. Be that way!"
    case _ => "Whatever."
  }

  private def isQuestion(statement: String) = statement.endsWith("?")
  private def isUpperCase(statement: String) = statement.toCharArray.exists(_.isUpper) && !statement.toCharArray.exists(_.isLower)
  private def isSilence(statement: String) = statement.isBlank
}

File: fsharp/bob/Bob.fs
module Bob

let (|Question|_|) (s: string) =
    if s.EndsWith('?') then
        Some Question
    else
        None

let (|Yelling|_|) (s: string) =
    if s.ToUpperInvariant() = s
       && String.exists System.Char.IsLetter s then
        Some Yelling
    else
        None

let response (input: string): string =
    match input.Trim() with
    | Yelling & Question -> "Calm down, I know what I'm doing!"
    | Yelling -> "Whoa, chill out!"
    | Question -> "Sure."
    | "" -> "Fine. Be that way!"
    | _ -> "Whatever."
